Source/Src_nd/ppm.F90

```diff
@@ -1355,2216 +1354,2219 @@ subroutine ppm_2d(lo,hi,s,s_lo,s_hi,nc_s,&
-
-          if (is_umac) then
-
-           ! u is MAC velocity -- use edge-based indexing
-                sigma = abs(u(i+1,j,k))*dt/dx(1)
-                s6 = SIX*s(i,j,k) - THREE*(sm+sp)
-                if (u(i+1,j,k) .gt. rel_eps) then
-                   Ip(i,j,k,1) = sp - &
-                        (sigma/TWO)*(sp-sm-(ONE-TWO3RD*sigma)*s6)
-                else
-                   Ip(i,j,k,1) = s(i,j,k)
+                sm = HALF*(s(i,j,k,n)+s(i-1,j,k,n)) - SIXTH*(dsvl_r-dsvl_l)
+                !
+                ! Make sure sedge lies in between adjacent cell-centered values.
+                !
+                sm = max(sm,min(s(i,j,k,n),s(i-1,j,k,n)))
+                sm = min(sm,max(s(i,j,k,n),s(i-1,j,k,n)))
+
+                ! sp
+                dsvl_l = ZERO
+                dsvl_r = ZERO
+
+                ! left side
+                dsc = HALF * (s(i+1,j,k,n) - s(i-1,j,k,n))
+                dsl = TWO  * (s(i,j,k,n) - s(i-1,j,k,n))
+                dsr = TWO  * (s(i+1,j,k,n) - s(i,j,k,n))
+                if (dsl*dsr .gt. ZERO) dsvl_l = sign(ONE,dsc)*min(abs(dsc),abs(dsl),abs(dsr))
+
+                ! right side
+                dsc = HALF * (s(i+2,j,k,n) - s(i,j,k,n))
+                dsl = TWO  * (s(i+1,j,k,n) - s(i,j,k,n))
+                dsr = TWO  * (s(i+2,j,k,n) - s(i+1,j,k,n))
+                if (dsl*dsr .gt. ZERO) dsvl_r = sign(ONE,dsc)*min(abs(dsc),abs(dsl),abs(dsr))
+
+                !
+                ! Interpolate s to x-edges.
+                !
+                sp = HALF*(s(i+1,j,k,n)+s(i,j,k,n)) - SIXTH*(dsvl_r-dsvl_l)
+                !
+                ! Make sure sedge lies in between adjacent cell-centered values.
+                !
+                sp = max(sp,min(s(i+1,j,k,n),s(i,j,k,n)))
+                sp = min(sp,max(s(i+1,j,k,n),s(i,j,k,n)))
+
+                !
+                ! Modify using quadratic limiters.
+                !
+                if ((sp-s(i,j,k,n))*(s(i,j,k,n)-sm) .le. ZERO) then
+                   sp = s(i,j,k,n)
+                   sm = s(i,j,k,n)
+                else if (abs(sp-s(i,j,k,n)) .ge. TWO*abs(sm-s(i,j,k,n))) then
+                   sp = THREE*s(i,j,k,n) - TWO*sm
+                else if (abs(sm-s(i,j,k,n)) .ge. TWO*abs(sp-s(i,j,k,n))) then
+                   sm = THREE*s(i,j,k,n) - TWO*sp
                 end if
 
-                sigma = abs(u(i,j,k))*dt/dx(1)
-                s6 = SIX*s(i,j,k) - THREE*(sm+sp)
-                if (u(i,j,k) .lt. -rel_eps) then
-                   Im(i,j,k,1) = sm + &
-                        (sigma/TWO)*(sp-sm+(ONE-TWO3RD*sigma)*s6)
-                else
-                   Im(i,j,k,1) = s(i,j,k)
+                ! Different stencil needed for x-component of EXT_DIR and HOEXTRAP adv_bc's.
+                !
+                if (i .eq. lo(1)+1 .and. lo(1)+1 .eq. domlo(1)) then
+                   if (adv_bc(1,1,bccomp) .eq. EXT_DIR  .or. adv_bc(1,1,bccomp) .eq. HOEXTRAP) then
```

Comment:
It's `lo(1)+1` because I've increased the size of the box passed in to include one ghost cell, so the new `lo(1)` is equal to the original `lo(1)-1`. But yes this would definitely be simpler reduced to a single if statement